src/libraries/System.Private.CoreLib/src/System/IO/PathInternal.cs

```diff
@@ -102,40 +102,52 @@ internal static bool AreRootsEqual(string? first, string? second, StringComparis
         }
 
         /// <summary>
-        /// Try to remove relative segments from the given path (without combining with a root).
+        /// Tries to remove relative segments from the given path, starting the analysis at the specified location.
         /// </summary>
-        /// <param name="path">Input path</param>
-        /// <param name="rootLength">The length of the root of the given path</param>
-        internal static string RemoveRelativeSegments(string path, int rootLength)
+        /// <param name="path">The input path.</param>
+        /// <param name="destination">The output path.</param>
+        /// <param name="charsWritten"></param>
+        /// <returns>The path without relative segments.</returns>
+        internal static bool TryRemoveRedundantSegments(ReadOnlySpan<char> path, Span<char> destination, out int charsWritten)
         {
-            var sb = new ValueStringBuilder(stackalloc char[260 /* PathInternal.MaxShortPath */]);
+            Debug.Assert(path.Length > 0);
 
-            if (RemoveRelativeSegments(path.AsSpan(), rootLength, ref sb))
-            {
-                path = sb.ToString();
-            }
+            charsWritten = 0;
+
+            ReadOnlySpan<char> pathRoot = Path.GetPathRoot(path);
+
+            // The resulting string is not expected to be larger than path
+            var sb = new ValueStringBuilder(destination);
```

Comment:
As mentioned, you may get better mileage by taking a `ValueStringBuilder` (by ref)  as your `destination` and you won't need `charsWritten` either.